### PR TITLE
feat(wasi): add support for python def and python module

### PR DESCRIPTION
**Pull request by afmika**
This solves MET-176 and MET-177

### CODE REVIEW

File: typegate/tests/runtimes/python_wasi/python_wasi_test.ts [hunk 1]
```diff
@@ -21,23 +21,44 @@ test("Python WASI runtime", async (t) => {
...
+    //       test: `test${i}`,
+    //     })
+    //     .on(e)
+    // );
+
+    // await Promise.all(tests);
+    // const end = performance.now();
     const duration = end - start;
-    assert(duration < 200, `Python WASI runtime was too slow: ${duration}ms`);
+    // from ~100ms (deno <-> wasi) to ~600ms (bindgen <-> wasmedge host <-> wasi)
```
Comment (zifeo) [inline typegate/tests/runtimes/python_wasi/python_wasi_test.ts:60]: Is not this just a start overhead of the vm?

**Review by zifeo — COMMENTED** [review]
Looks good!

File: typegate/native/src/runtimes/python/wasi_vm.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        .register_module(None, module)?;
+
+    let wasi_module = vm.wasi_module_mut().unwrap();
+
+    let mut preopens = vec![format!(
+        "/usr/local/lib:{}:readonly",
+        pythonlib_path.display()
+    )];
+    preopens.extend(inp_preopens);
+    let preopens = preopens.iter().map(|s| &s[..]).collect();
```
Comment (zifeo) [inline typegate/native/src/runtimes/python/wasi_vm.rs:52]: is not there a "cleaner" way to transform like into or as_ref?

File: typegate/native/src/runtimes/python/wasi_vm.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    wasi_module.initialize(None, None, Some(preopens));
+
+    println!("wasi_module: {:?}", wasi_module.exit_code());
```
Comment (zifeo) [inline typegate/native/src/runtimes/python/wasi_vm.rs:55]: this should always 0, I would suggest to throw if not (with some error details)?

File: typegate/native/src/runtimes/python/wasi_vm.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    // if wasi-vfs is not used, initialize the reactor as not done automatically
+    let _init = vm.run_func(None, "_initialize", params!())?;
```
Comment (zifeo) [inline typegate/native/src/runtimes/python/wasi_vm.rs:58]: check init return code also?

File: typegate/native/src/runtimes/python/wasi_vm.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+use std::path::PathBuf;
+use wasmedge_sdk::{
+    config::{CommonConfigOptions, ConfigBuilder, HostRegistrationConfigOptions},
+    error::HostFuncError,
+    host_function, params, Caller, ImportObjectBuilder, Module, Vm, VmBuilder, WasmValue,
+};
+
+#[host_function]
+pub fn callback(_caller: Caller, _args: Vec<WasmValue>) -> Result<Vec<WasmValue>, HostFuncError> {
+    // println!("[host] callback");
```
Comment (zifeo) [inline typegate/native/src/runtimes/python/wasi_vm.rs:13]: fyi & async step, the callback will have to be used once Tokio is used within runtime

File: typegate/native/Cargo.toml [hunk 6]
```diff
@@ -50,7 +50,11 @@ rust-s3 = "0.33.0"
 http = "0.2.9"
 envconfig = "0.10.0"
 once_cell = "1.18.0"
-wasmedge-sdk = { version = "0.8.1", features = ["standalone"] }
 base64 = "0.21.2"
 tap = "1.0.1"
+num-traits = "0.2"
+num-derive = "0.3"
+wasmedge-sdk = { version = "0.8.1", features = ["standalone"] }
```
Comment (zifeo) [inline typegate/native/Cargo.toml:57]: can you add a comment here why 0.9.0 did not work?

**Review by zifeo — APPROVED** [review]

File: typegate/native/Cargo.toml (hunk 6 above)
Comment (afmika) [reply typegate/native/Cargo.toml:57]: 1. Vm -> Vm<T>: we have to update wasmedge_sdk_bindgen
2. After some cleanups, even loading the reactor simply did not work (no panic, nothing) the program just quit with exit code 11.

File: typegate/native/src/runtimes/python/wasi_vm.rs (hunk 4 above)
Comment (afmika) [reply typegate/native/src/runtimes/python/wasi_vm.rs:58]: yup, good idea 👍

File: typegate/native/src/runtimes/python/wasi_vm.rs (hunk 2 above)
Comment (afmika) [reply typegate/native/src/runtimes/python/wasi_vm.rs:52]: `as_ref` should work

File: typegate/src/runtimes/python_wasi/python_vm.ts [hunk 7]
```diff
@@ -46,6 +54,28 @@ function promisify<T>(fn: CallableFunction): Promise<T> {
   });
 }
 
+// TODO: update to new version
+const pythonWasiReactorUrl =
+  "https://github.com/metatypedev/python-wasi-reactor/releases/download/v0.1.0/python3.11.1-wasi-reactor.wasm.tar.gz";
+const cachePathReactor = join(config.tmp_dir, "python3.11.1-wasi-reactor.wasm");
+
+const libPythonUrl =
+  "https://github.com/assambar/webassembly-language-runtimes/releases/download/python%2F3.11.1%2B20230223-8a6223c/libpython-3.11.1.tar.gz";
```
Comment (zifeo) [inline typegate/src/runtimes/python_wasi/python_vm.ts:63]: https://github.com/vmware-labs/webassembly-language-runtimes/releases/download/python%2F3.11.4%2B20230714-11be424/libpython-3.11.4-wasi-sdk-20.0.tar.gz is the new one

File: typegate/src/runtimes/python_wasi/python_vm.ts [hunk 8]
```diff
@@ -46,6 +54,28 @@ function promisify<T>(fn: CallableFunction): Promise<T> {
   });
 }
 
+// TODO: update to new version
+const pythonWasiReactorUrl =
```
Comment (zifeo) [inline typegate/src/runtimes/python_wasi/python_vm.ts:58]: https://github.com/metatypedev/python-wasi-reactor/releases/tag/v0.2.0

File: typegate/native/src/runtimes/python/wasi_vm.rs (hunk 4 above)
Comment (afmika) [reply typegate/native/src/runtimes/python/wasi_vm.rs:58]: Actually, it's just returning an empty vector.

**Review by zifeo — DISMISSED** [review]

File: installer.sh [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+fi
+
+tar -xzf "$ASSET.$EXT" "$EXE"
+rm "$ASSET.$EXT"
+chmod +x "$EXE"
+
+cat <<EOF
+
+$NAME has been extracted to your folder, please move it to /usr/local/bin (or any other folder on your PATH):
+$ sudo mv $EXE /usr/local/bin
```
[the remark this answers is not in the document]
Comment (zifeo) [reply installer.sh:89]: Good points, will fix in subsequent PR

**Pull request merged**